app/src/main/java/net/gini/android/gvlexample/GVLExampleActivity.java

```diff
@@ -0,0 +1,359 @@
+                            }
+                        })
+                .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(final DialogInterface dialog, final int which) {
+                        listener.permissionDenied();
+                    }
+                })
+                .setOnCancelListener(new DialogInterface.OnCancelListener() {
+                    @Override
+                    public void onCancel(final DialogInterface dialog) {
+                        listener.permissionDenied();
+                    }
+                })
+                .create();
+        alertDialog.show();
+    }
+
+    private void showStoragePermissionRationale(final PermissionToken token) {
+        final AlertDialog alertDialog = new AlertDialog.Builder(this)
+                .setMessage("To analyze images and PDFs from the device storage access is needed.")
+                .setPositiveButton("Grant Access",
+                        new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(final DialogInterface dialog, final int which) {
+                                token.continuePermissionRequest();
+                            }
+                        })
+                .setOnCancelListener(new DialogInterface.OnCancelListener() {
+                    @Override
+                    public void onCancel(final DialogInterface dialog) {
+                        token.cancelPermissionRequest();
+                    }
+                })
+                .create();
+        alertDialog.show();
+    }
+
+    private void showCameraPermissionDeniedDialog(final PermissionRequestListener listener) {
+        final AlertDialog alertDialog = new AlertDialog.Builder(this)
+                .setMessage(
+                        "You need to grant access to the camera in order to use the Gini Vision Library.")
+                .setPositiveButton(
+                        "Grant Access", new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(final DialogInterface dialog, final int which) {
+                                showAppDetailsSettingsScreen();
+                            }
+                        })
+                .setNegativeButton("Cancel", null)
+                .setOnCancelListener(new DialogInterface.OnCancelListener() {
+                    @Override
+                    public void onCancel(final DialogInterface dialog) {
+                        listener.permissionDenied();
+                    }
+                })
+                .create();
+        alertDialog.show();
+    }
+
+    private void showAppDetailsSettingsScreen() {
+        final Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
+        final Uri uri = Uri.fromParts("package", getPackageName(), null);
```

Comment:
Does Android provide a way to get the _scheme_ instead of hardcoding it here?

Reply:
Nope, I only found constants for content, file and android resource in the [ContentResolver](https://developer.android.com/reference/android/content/ContentResolver.html)... 😢